perf_bench_test.go

```diff
@@ -24,127 +24,619 @@ package main
+		"+ to add, "+
+		"-|^ to remove, "+
+		"comma separated list, wildcards allowed, "+
+		"add * to include all children/siblings, "+
+		"add /level to specify highest level node to match, "+
+		"add ^ to match related siblings and metrics, "+
+		"start with ! to only include specified nodes)")
 )
 
-func TestBenchRequestPerSecond(t *testing.T) {
+// TestProfileIncrementConfiguration loads requests to VMs and increments VM number after loads start to violate time latency constraint.
+// It also profile counters and RPS at each step. After iteration finishes, it saves results in bench.csv under benchDir folder and
+// plots each counters which are also saved under benchDir folder
+func TestProfileIncrementConfiguration(t *testing.T) {
+	var (
+		idx, rps      int
+		pinnedFuncNum int
+		startVMID     int
+		servedTh      uint64
+		isSyncOffload bool = true
+		metrFile           = "bench.csv"
+		images             = getImages(t)
+		cores              = runtime.NumCPU()
+		metrics            = make([]map[string]float64, *maxVMNum / *vmIncrStep)
+	)
+	log.SetLevel(log.InfoLevel)
+
+	validateRuntimeArguments(t)
+
+	createResultsDir()
+
+	funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
+
+	for vmNum := *vmIncrStep; vmNum <= *maxVMNum; vmNum += *vmIncrStep {
+		if vmNum < cores {
+			rps = calculateRPS(vmNum)
+		} else {
+			rps = calculateRPS(cores)
+		}
+
+		log.Infof("vmNum: %d, Target RPS: %d", vmNum, rps)
+
+		bootVMs(t, images, startVMID, vmNum)
+		metrics[idx] = loadAndProfile(t, images, vmNum, rps, isSyncOffload)
+		startVMID = vmNum
+		idx++
+	}
+
+	dumpMetrics(t, metrics, metrFile)
+	profile.PlotCVS(*vmIncrStep, *benchDir, metrFile, "the number of VM")
+
+	tearDownVMs(t, images, startVMID, isSyncOffload)
+}
+
+// TestProfileSingleConfiguration loads requests to fixed number of VMs until loads start to violate tail latency constraint
+// and then saves the results in bench.csv under benchDir folder
+func TestProfileSingleConfiguration(t *testing.T) {
 	var (
 		servedTh      uint64
 		pinnedFuncNum int
-		vmID          int
 		isSyncOffload bool = true
-		images             = getImages()
-		funcs              = []string{}
-		timeInterval       = time.Duration(time.Second.Nanoseconds() / int64(*targetReqPerSec))
-		totalRequests      = *executionTime * *targetReqPerSec
+		images             = getImages(t)
 	)
 
 	log.SetLevel(log.InfoLevel)
 
+	validateRuntimeArguments(t)
+
+	createResultsDir()
+
 	funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
 
+	bootVMs(t, images, 0, *vmNum)
+
+	serveMetrics := loadAndProfile(t, images, *vmNum, *targetRPS, isSyncOffload)
+
+	tearDownVMs(t, images, *vmNum, isSyncOffload)
+
+	dumpMetrics(t, []map[string]float64{serveMetrics}, "bench.csv")
+}
+
+// TestColocateVMsOnSameCPU measures the differences between 2 VMs on the same core and 2VMs on different cores
+// Only works for a 2x 16-core machine or above
```

Comment:
please re-phrase as I wrote in the manual.

Reply:
Done